## <p style="text-align: right;"> &#9989; Put your name here</p>

# CMSE 201 Midterm (Section 001 - Fall 2021)

The exam is set up so that even if you get stuck on one part there are opportunities to get points on the other parts, so consider jumping ahead if you feel like you aren't making progress and then come back later if you have time.

**Important note about using online resources**: This exam is "open internet". That means that you can look up documentation, google how to accomplish certain Python tasks, access old notebooks, etc. **However, the use of any person-to-person communication software is absolutely not acceptable.** If you are seen accessing your email, using a chat program (e.g. Slack), or any sort of collaborative cloud storage or document software (e.g. Google Documents), you will be at risk for receiving a zero on the exam. 

**Keep your eyes on your screen!** Unfortunately, there isn't enough space in the room for everyone to sit at their own table so please do your best to keep your eyes on your own screen. This exam is designed to give *you* the opportunity to show the instructor what you can do and you should hold yourself accountable for maintaining a high level of academic integrity. If any of the instructors observe suspicious behavior, you will, again, risk receiving a zero.

___
## Overview:

___
## Part 0: Academic Honesty Statement (3 points)

**Insert your personalized academic honesty statement here.**

*your statement here*

## Analyzing Data from Metropolitan Statistical Areas

Metropolitan Statistical Areas (MSAs) are geographical regions of high population densities that usually encompass multiple cities. Here in Michigan, the area encompassing Lansing and East Lansing is considered an MSA, as well as Detroit/Warren/Dearborn. 

For this exam, you will be working with data from the 30 most populous MSAs in the US. The data has been provided below, as four lists. These lists are:
-The name of the MSA (`msa_name`)
-The population of the MSA in 2020 (`msa_pop_2020`)
-The population of the MSA in 2010 (`msa_pop_2010`)
-The region the MSA is located in (East, Midwest, South, or West; `msa_region`)

### Your Job
You work for a company that does real estate development. They are planning a large new block of condominiums, but they’re still not certain of a location for the project. **Your job is to analyze the MSA data and determine the fastest growing region.**


In [1]:
msa_name =['New York City','Los Angeles','Chicago','Dallas','Houston','Washington DC','Philadelphia','Miami','Atlanta','Boston','Phoenix','San Francisco','Riverside','Detroit','Seattle','Twin Cities','San Diego','Tampa','Denver','Baltimore','St. Louis','Orlando','Charlotte','San Antonio','Portland','Sacramento','Pittsburgh','Austin','Las Vegas','Cincinnati']
msa_pop_2020 = [20140470,13200998,9618502,7637387,7122240,6385162,6245051,6138333,6089815,4941632,4845832,4749008,4599839,4392041,4018762,3690261,3298634,3175275,2963821,2844510,2820253,2673376,2660329,2558143,2512859,2397382,2370930,2283371,2265461,2256884]
msa_pop_2010 = [18897109,12828837,9461105,6366542,5920416,5649540,5965343,5564635,5286728,4552402,4192887,4335391,4224851,4296250,3439809,3346859,3095313,2783243,2543482,2710489,2787701,2134411,2243960,2142508,2226009,2149127,2356285,1716289,1951269,2137667]
msa_region = ['East','West','Midwest','South','South','South','East','South','South','East','West','West','West','Midwest','West','Midwest','West','South','West','East','Midwest','South','South','South','West','West','East','South','West','Midwest']

### Part 1: Calculating the Growth Rate of MSAs

First, calculate the growth rate of the different MSAs over the last ten years. The equation for calculating growth rate is:

\begin{equation}
\mathrm{Growth~Rate} = \frac{\mathrm{2020~Population} - \mathrm{2010~Population}}{\mathrm{2010~Population}}
\end{equation}

So, for example, the growth rate of New York would be:
\begin{equation}
\mathrm{NY Growth~Rate} = \frac{20140470 - 18897109}{18897109} = 0.0658
\end{equation}
Or 6.58%.

**Do this for all MSAs and append all growth rate values to a new list, titled `msa_growth`.**

In [ ]:
#Write your code here

In [2]:
###ANSWER
msa_growth= []
for jj in range(len(msa_pop_2020)):
    growth_rate = (msa_pop_2020[jj]-msa_pop_2010[jj])/msa_pop_2010[jj]
    msa_growth.append(growth_rate)

### Part 2: Organizing your Data

Next, create a list of lists for the MSAs. A list for a single MSA should have the following information:

```
[msa_name, msa_pop_2020, msa_pop_2010, msa_region, msa_growth]
```

So, for example, the list for New York City would be:
```
['New York City', 20140470, 18897109, 'East', 0.0658]
```

**Create lists for each MSA. When you have created the list for an MSA, append that list to a list called `all_msa_info`.**

In [ ]:
#Write your code here

In [3]:
###ANSWER
all_msa_info = []

for jj in range(len(msa_pop_2020)):
    temp_list = [msa_name[jj],msa_pop_2020[jj],msa_pop_2010[jj],msa_region[jj],msa_growth[jj]]
    all_msa_info.append(temp_list)

### Part 3: Selecting MSAs by Region
Next, create a function that takes in your `all_msa_info` and a region, and returns all of the MSA lists that are in that region. So, for example, if I passed this function `all_msa_info` and `'East'`, it would return a list that looked like:

```
[['New York City', 20140470, 18897109, 'East', 0.0658],
['Philidelphia',...]
etc.
```

In [ ]:
#Write your code here

In [4]:
###ANSWER

def get_regional_msas(msa_list,region):
    regional_msa_list = []
    for jj in range(len(msa_list)):
        if msa_list[jj][3] == region:
            regional_msa_list.append(msa_list[jj])
    return regional_msa_list

### 3.1 STOP

**Test your function to make sure that it is working correctly. Use it to create a list of all MSAs in the Midwest.**

In [ ]:
#Write your code here

In [5]:
###ANSWER

midwest_msas = get_regional_msas(all_msa_info,'Midwest')
for kk in range(len(midwest_msas)):
    print(midwest_msas[kk])

['Chicago', 9618502, 9461105, 'Midwest', 0.016636217439717667]
['Detroit', 4392041, 4296250, 'Midwest', 0.022296421297643292]
['Twin Cities', 3690261, 3346859, 'Midwest', 0.10260426268330994]
['St. Louis', 2820253, 2787701, 'Midwest', 0.011677005532515862]
['Cincinnati', 2256884, 2137667, 'Midwest', 0.05576967787779855]


### Part 4: Fixing a Function to Select Data by Index

Below is a function that takes two values: a list of MSA lists (in the format of `all_msa_info`) and an index/column value (an integer). The function is supposed to return a list of all values from the list of MSA lists in that index position.  So, for instance, if we passed the function `all_msa_info` and `0`, it is supposed to return:

```
['New York City','Los Angeles','Chicago',...]
```

This is because the index `0` corresponds to the name of the MSA, and `all_msa_info` contains all MSAs. However, the code isn't currently working.

**Fix the function below so that it does what it's supposed to do.**

In [ ]:
def get_data_at_index(msa_list,index_val):
    data_at_index = []
    for jj in range(len(msa_list)):
        temp_value = msa_list[jj][5]
    return data_at_index
get_data_at_index(all_msa_info,0)

In [6]:
###ANSWER

def get_data_at_index(msa_list,index_val):
    data_at_index = []
    for jj in range(len(msa_list)):
        temp_value = msa_list[jj][index_val]
        data_at_index.append(temp_value)
    return data_at_index

### 4.1 STOP

Test your function to make sure that it is working correctly. Use it to create a list of growth rates for the MSAs in the Midwest.

In [ ]:
#Write your code here

In [ ]:
###ANSWER

midwest_growth_rates = get_data_at_index(midwest_msas,4)
print(midwest_growth_rates)

### Part 5: Calculating Average Growth Rate

Use the two functions you’ve defined above to:
1. Select MSA data for each of the four regions 
2. Get lists of growth rates for each of the four regions 
3. Calculate and print out the average (mean) growth rate for the region

In [9]:
###ANSWER
import numpy as np

regions = ["East","West","Midwest","South"]
for region in regions:
    regional_msa = get_regional_msas(all_msa_info,region)
    regional_gr = get_data_at_index(regional_msa,4)
    print(region," region has had a mean growth rate of ",np.mean(np.array(regional_gr)))


East  region has had a mean growth rate of  0.050769146146288927
West  region has had a mean growth rate of  0.11735531996871829
Midwest  region has had a mean growth rate of  0.04179671696619706
South  region has had a mean growth rate of  0.18911461913715216


### Part 6: Summarizing your Results
If your company wants to build where there’s fast growth, what region should your company put the new construction project in? 

*Write your answer here*

### ANSWER

They should build in the south, which has the fastest growing metros (an average of nearly 19%)

---
## You're done! Congrats on finishing your CMSE 201 Midterm!

Upload a copy of this notebook to the submission folder on D2L and make sure your name is on it!